File: with-lombok/src/test/java/TestData.java
import static org.assertj.core.api.Assertions.*;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import lombok.Data;
import lombok.Value;

class TestData {

    /**
     * Mutable Java bean with toString, equal and hashcode
     */
    @Data
    static class UserData {
        private String firstname;
        private String lastName;
    }

    @Value
    static class UserValue {
        String firstname;
        String lastname;
    }

    @Test
    @DisplayName("should data classes implement toString")
    void shouldDataImplementToString() {
        UserData userData = new UserData();

        userData.setFirstname("John");
        userData.setLastName("Doe");

        assertThat(userData).hasToString("TestData.UserData(firstname=John, lastName=Doe)");
    }

    @Test
    @DisplayName("should value classes implement toString")
    void shouldValueImplementToString(){
        UserValue userValue = new UserValue("Jane", "Doe");

        assertThat(userValue).hasToString("TestData.UserValue(firstname=Jane, lastname=Doe)");
    }
}

File: with-kotlin/src/test/kotlin/TestData.kt
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class TestData {
    /**
     * Mutable Data class with nullable attributes with default value
     */
    data class UserData(
        var firstname: String? = null,
        var lastname: String? = null,
    )

    /**
     * Immutable Data class
     */
    data class UserDataImmutable(val firstname: String?, val lastname: String?)

    @Test
    fun `should data class implement toString`() {
        val user1 = UserData()
        user1.firstname = "John"
        user1.lastname = "Doe"

        assertEquals("UserData(firstname=John, lastname=Doe)", user1.toString())
    }

    @Test
    fun `should data class implement equals`() {
        val user1 = UserData("John", "Doe")
        val user2 = UserData("John", "Doe")

        assertEquals(user1, user2)
    }

    @Test
    fun `should create an immutable user`(){
        val immutableUser = UserDataImmutable("François","Pignon")
//        immutableUser.firstname = "Jean"

        assertEquals("UserDataImmutable(firstname=François, lastname=Pignon)",immutableUser.toString())
    }
}
